### PR TITLE
DRILL-6647: Update Calcite version to 1.17.0

### PR DESCRIPTION
**Pull request by vvysotskyi**
Most of the changes are connected with changes in Avatica (https://github.com/apache/calcite-avatica/commit/2ebfd8060119efde140a0343fbf9e833b2915505) where was added a check for resource is closed, so most of the methods in Drill with this check became redundant and were removed.

For more details please see [DRILL-6647](https://issues.apache.org/jira/browse/DRILL-6647)

### CODE REVIEW

**Comment by priteshm** [comment]
@HanumathRao can you review this?

**Comment by vvysotskyi** [comment]
Rebased onto master. 
@HanumathRao, did you have a chance to take a look at this PR?

File: exec/java-exec/src/main/codegen/data/Parser.tdd [hunk 1]
```diff
@@ -85,13 +85,17 @@
...
-    dropStatementParserMethods: [
+  dropStatementParserMethods: [
   ]
 
   # List of keywords from "keywords" section that are not reserved.
-    nonReservedKeywords: [
+  nonReservedKeywords: [
   ]
 
+  # List of additional join types. Each is a method with no arguments.
```
Comment (HanumathRao) [inline exec/java-exec/src/main/codegen/data/Parser.tdd]: I am not clear as to why are these changes required for the upgrade?

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillConditions.java [hunk 2]
```diff
@@ -19,24 +19,14 @@
...
-    }
-
-    @Override
-    public boolean test(RexNode expr) {
-      if (expr instanceof RexCall) {
-        RexCall call = (RexCall)expr;
-        return "item".equals(call.getOperator().getName().toLowerCase());
-      }
-      return false;
+  public static PushProjector.ExprCondition PRESERVE_ITEM = expr -> {
```
Comment (HanumathRao) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillConditions.java]: Should this also be made final ??

File: exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillConvertletTable.java [hunk 3]
```diff
@@ -36,19 +37,36 @@
...
-      return cx.getRexBuilder().makeCall(SqlStdOperatorTable.SQRT, operand);
+
+  private static final SqlRexConvertlet SQRT_CONVERTLET = (cx, call) -> {
+    RexNode operand = cx.convertExpression(call.operand(0));
+    return cx.getRexBuilder().makeCall(SqlStdOperatorTable.SQRT, operand);
+  };
+
+  // Rewrites COALESCE function into CASE WHEN IS NOT NULL operand1 THEN operand1...
+  private static final SqlRexConvertlet COALESCE_CONVERTLET = (cx, call) -> {
+    if (call.operandCount() == 1) {
```
Comment (HanumathRao) [inline exec/java-exec/src/main/java/org/apache/drill/exec/planner/sql/DrillConvertletTable.java]: can you create a local variable something like argCount for call.operandCount() and use it.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/AbstractSchema.java [hunk 4]
```diff
@@ -161,6 +164,30 @@ public boolean showInInformationSchema() {
...
+   * Override this method to change their behavior.</p>
+   *
+   * @return Map of types in this schema by name
+   */
+  protected Map<String, RelProtoDataType> getTypeMap() {
+    return ImmutableMap.of();
+  }
+
+  @Override
+  public Set<String> getTypeNames() {
```
Comment (HanumathRao) [inline exec/java-exec/src/main/java/org/apache/drill/exec/store/AbstractSchema.java]: Are these new functions used in the code?

**Comment by HanumathRao** [comment]
@vvysotskyi  Thank you for making these changes. Changes look good to me.

File: exec/java-exec/src/main/java/org/apache/drill/exec/store/AbstractSchema.java (hunk 4 above)
Comment (vvysotskyi) [reply exec/java-exec/src/main/java/org/apache/drill/exec/store/AbstractSchema.java]: In Drill code, they are not used but in Calcite in some places, but it only affects functionality that is not used by Drill yet (CALCITE-2045). Introducing them is required since these methods are added to the `Schema` interface.

File: exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillResultSetImpl.java [hunk 5]
```diff
@@ -162,1713 +160,1097 @@ public void close() {
...
-  }
-
-  //----------------------------------------------------------------
-  @Override
-  public int findColumn( String columnLabel ) throws SQLException {
-    throwIfClosed();
-    return super.findColumn( columnLabel );
-  }
-
-  //--------------------------JDBC 2.0-----------------------------------
```
[the remark this answers is not in the document]
Comment (vvysotskyi) [reply exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillResultSetImpl.java]: Thanks, moved it back. We don't need a section for Getters and Setters.

File: exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillStatementImpl.java [hunk 6]
```diff
@@ -65,107 +65,52 @@ private void throwIfClosed() throws AlreadyClosedSqlException {
...
+  public DrillConnectionImpl getConnection() throws SQLException {
+    checkOpen();
     return connection;
   }
 
-  // WORKAROUND:  Work around AvaticaStatement's code that wraps _any_ exception,
-  // even if SQLException, by unwrapping to get cause exception so caller can
-  // throw it directly if it's a SQLException:
-  // TODO:  Any ideas for a better name?
-  private SQLException unwrapIfExtra( final SQLException superMethodException ) {
```
[the remark this answers is not in the document]
Comment (vvysotskyi) [reply exec/jdbc/src/main/java/org/apache/drill/jdbc/impl/DrillStatementImpl.java]: Exactly this method is not required since now we don't wrap `SQLException` exceptions into `SQLException`. Other methods, as I mentioned in the previous comment, methods in Avatica have a check for connection is open.

File: exec/jdbc/src/test/java/org/apache/drill/jdbc/JdbcTestBase.java [hunk 7]
```diff
@@ -278,27 +272,17 @@ public static String toString(ResultSet resultSet) throws SQLException {
...
+      this.adapter = () -> factory.getConnection(
+          new ConnectionInfo("jdbc:drill:zk=local", ModelAndSchema.this.info));
     }
 
     public TestDataConnection sql(String sql) {
       return new TestDataConnection(adapter, sql);
     }
 
     public <T> T withConnection(Function<Connection, T> function) throws Exception {
-      Connection connection = null;
```
[the remark this answers is not in the document]
Comment (vvysotskyi) [reply exec/jdbc/src/test/java/org/apache/drill/jdbc/JdbcTestBase.java]: It is closed automatically since `try with resources` is used.

**Comment by HanumathRao** [comment]
@vvysotskyi  Changes look good to me.

File: pom.xml [hunk 8]
```diff
@@ -45,8 +45,8 @@
     <dep.guava.version>18.0</dep.guava.version>
     <forkCount>2</forkCount>
     <parquet.version>1.10.0</parquet.version>
-    <calcite.version>1.16.0-drill-r7</calcite.version>
-    <avatica.version>1.11.0</avatica.version>
+    <calcite.version>1.17.0-drill-r0</calcite.version>
```
Comment (vrozov) [inline pom.xml]: What is the plan to start publishing Apache project artifacts to Apache maven repository instead of a proprietary one?

**Comment by gparai** [comment]
@vvysotskyi thanks for addressing my comments. LGTM +1

File: pom.xml (hunk 8 above)
Comment (vvysotskyi) [reply pom.xml]: I don't know the correct answer and how such cases of specific versions are handled in other Apache projects, so I may be wrong with my assumptions. 

Since this version contains Drill-specific commits and is used by Drill only, I'm not sure that it will be justified to publish it to the Apache maven repository.

Comment (vvysotskyi) [reply pom.xml]: @vrozov, just to clarify for batch committers: Is this question blocks merging of this PR?

Comment (vrozov) [reply pom.xml]: Note for batch committer: no, the question is not a blocker for merging the PR.

@vvysotskyi I don't see why other projects will be confused by publishing an artifact that contains "drill" classifier. It sounds strange to me that an Apache project consumes an artifact produced by another Apache project not from the Apache repository.

Comment (vvysotskyi) [reply pom.xml]: This version is built using the sources taken from our fork of Apache Calcite.

My expectation was if Apache maven repository contains some version of the project, then I can find the sources which were used for building the artifacts in the corresponding Apache repository.

Comment (vrozov) [reply pom.xml]: I don't know where the expectation for the source code in github comes from especially that github is not part of Apache INFRA, github Apache repositories are actually forks from Apache git and not all Apache projects use git as scm.

To clarify, it does not matter to me whether the artifact is published to Apache maven repository or another repository as long as it is published to a repository that is mirrored by maven central and the artifact is searchable in https://search.maven.org/

Comment (arina-ielchiieva) [reply pom.xml]: @vrozov thanks for bringing up this point. Never thought about it but definitely valid one.
I have created Jira - https://issues.apache.org/jira/browse/DRILL-6711
